pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,430 @@
+		res := &CachingBucket{}
+		*res = *cb
+		res.bucket = ib.WithExpectedErrs(expectedFunc)
+		return res
+	}
+
+	return cb
+}
+
+func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailureExpectedFunc) objstore.BucketReader {
+	return cb.WithExpectedErrs(expectedFunc)
+}
+
+func (cb *CachingBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	return cb.bucket.Iter(ctx, dir, f)
+}
+
+func (cb *CachingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return cb.bucket.Get(ctx, name)
+}
+
+var chunksMatcher = regexp.MustCompile(`^.*/chunks/\d+$`)
+
+func isTSDBChunkFile(name string) bool {
+	return chunksMatcher.MatchString(name)
+}
+
+func (cb *CachingBucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	if isTSDBChunkFile(name) && off >= 0 && length > 0 {
+		return cb.getRangeChunkFile(ctx, name, off, length)
+	}
+
+	return cb.bucket.GetRange(ctx, name, off, length)
+}
+
+func (cb *CachingBucket) Exists(ctx context.Context, name string) (bool, error) {
+	return cb.bucket.Exists(ctx, name)
+}
+
+func (cb *CachingBucket) IsObjNotFoundErr(err error) bool {
+	return cb.bucket.IsObjNotFoundErr(err)
+}
+
+func (cb *CachingBucket) ObjectSize(ctx context.Context, name string) (uint64, error) {
+	return cb.bucket.ObjectSize(ctx, name)
+}
+
+func (cb *CachingBucket) cachedObjectSize(ctx context.Context, name string, ttl time.Duration) (uint64, error) {
+	key := cachingKeyObjectSize(name)
+
+	cb.objectSizeRequests.Add(1)
+
+	hits := cb.cache.Fetch(ctx, []string{key})
+	if s := hits[key]; len(s) == 8 {
+		cb.objectSizeHits.Add(1)
+		return binary.BigEndian.Uint64(s), nil
+	}
+
+	size, err := cb.bucket.ObjectSize(ctx, name)
+	if err != nil {
+		return 0, err
+	}
+
+	var buf [8]byte
+	binary.BigEndian.PutUint64(buf[:], size)
+	cb.cache.Store(ctx, map[string][]byte{key: buf[:]}, ttl)
+
+	return size, nil
+}
+
+func (cb *CachingBucket) getRangeChunkFile(ctx context.Context, name string, offset, length int64) (io.ReadCloser, error) {
+	cb.requestedChunkBytes.Add(float64(length))
+
+	size, err := cb.cachedObjectSize(ctx, name, cb.config.ChunkObjectSizeTTL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get size of chunk file: %s", name)
+	}
+
+	// If length goes over object size, adjust length. We use it later to limit number of read bytes.
+	if uint64(offset+length) > size {
+		length = int64(size - uint64(offset))
+	}
+
+	// Start and end range are block-aligned offsets into object, that we're going to read.
+	startRange := (offset / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+	endRange := ((offset + length) / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+	if (offset+length)%cb.config.ChunkBlockSize > 0 {
+		endRange += cb.config.ChunkBlockSize
+	}
+
+	// The very last block in the object may have length that is not divisible by block size.
+	lastBlockOffset := int64(-1)
```

Comment:
Despite correct (due to how the current logic works) setting it to the correct offset may be more future proof. I think you can easily set it to `endRange - cb.config.ChunkBlockSize` thus covering all cases.